test/e2e/upgrade_test.go

```diff
@@ -31,15 +31,71 @@ const (
+func getNodeAffinities() ([]*v1.Affinity, error) {
+	var affinities []*v1.Affinity
+	if gc.numberOfMachineNodes >= 1 {
+		affinityNoByoh, err := getNodeAffinityForLabel(v1.NodeSelectorOpDoesNotExist, controllers.BYOHLabel)
+		if err != nil {
+			return nil, errors.Wrapf(err, "error creating node affinity without BYOH label")
+		}
+		affinities = append(affinities, affinityNoByoh)
+	}
+	if gc.numberOfBYOHNodes >= 1 {
+		affinityNoByoh, err := getNodeAffinityForLabel(v1.NodeSelectorOpIn, controllers.BYOHLabel, "true")
+		if err != nil {
+			return nil, errors.Wrapf(err, "error creating node affinity for BYOH label")
+		}
+		affinities = append(affinities, affinityNoByoh)
+	}
+	return affinities, nil
+}
+
+// getLinuxJobCommand creates a Job to curl Windows webserver every 5 seconds.
+// If the webserver is not accessible for few minutes Curl Windows webserver
+// exit the pod with an error.
+func getLinuxJobCommand(clusterIP string) []string {
+	return []string{
+		"bash",
+		"-c",
+		" while true;" +
+			" do curl " + clusterIP + ";" +
+			" if [ $? != 0 ]; then" +
+			" sleep 60;" +
+			" curl " + clusterIP + " || exit 1;" +
+			" fi; " +
+			"sleep 5; " +
+			"done"}
+}
+
 // upgradeTestSuite tests behaviour of the operator when an upgrade takes place.
 func upgradeTestSuite(t *testing.T) {
 	testCtx, err := NewTestContext()
 	require.NoError(t, err)
 
-	// test if Windows workloads are running by creating a Job that curls the workloads continuously.
-	testerJob, err := testCtx.deployWindowsWorkloadAndTester()
-	require.NoError(t, err, "error testing Windows workloads")
-	defer testCtx.deleteJob(testerJob.Name)
+	// generate node affinities
+	affinities, err := getNodeAffinities()
+	require.NoError(t, err, "error getting node affinity")
+	// loop affinities
+	for _, a := range affinities {
+		// wraps into an inline function to ensure the right execution order of deferred calls
+		func(affinity *v1.Affinity) {
+			// test if Windows workloads are running by creating a Job that curls the workloads continuously.
+			deployment, err := testCtx.deployWindowsWebServer("win-webserver", affinity)
+			require.NoErrorf(t, err, "error creating Windows Webserver deployment for upgrade test")
+			defer testCtx.deleteDeployment(deployment.Name)
+
+			// create a clusterIP service which can be used to reach the Windows webserver
+			intermediarySVC, err := testCtx.createService(deployment.Name, v1.ServiceTypeClusterIP, *deployment.Spec.Selector)
+			require.NoErrorf(t, err, "error creating service for deployment %s", deployment.Name)
+			defer testCtx.deleteService(intermediarySVC.Name)
+
+			// create a Job object that continuously curls the webserver every 5 seconds.
+			testerJob, err := testCtx.createLinuxJob(windowsWorkloadTesterJob, getLinuxJobCommand(intermediarySVC.Spec.ClusterIP))
+			require.NoErrorf(t, err, "error creating linux job %s", windowsWorkloadTesterJob)
+			defer testCtx.deleteJob(testerJob.Name)
```

Comment:
This snippet could be a helper function.
I am also a little confused by this function, are the defer calls executed at the end of entire suite? Just making sure this implementation is solving the problem of not deleting deployments before running upgrade tests.

Reply:
Moving the loop content to a helper is similar to what it's been [removed](https://github.com/openshift/windows-machine-config-operator/pull/645/commits/3551877c5aca09efea84533bb131bb105505ff78#diff-fc8ea8c7c1a732fd9a425771cb3463d1cd143e5527acec751f32ba88f2df697aL157), in order to allow the deferred calls to execute in the current context, so the deployments do not get deleted before the upgrade process starts.

Reply:
> I am also a little confused by this function, are the defer calls executed at the end of entire suite?

No, they are executed just before [upgradeTestSuite](https://github.com/openshift/windows-machine-config-operator/pull/645/commits/3551877c5aca09efea84533bb131bb105505ff78#diff-fc8ea8c7c1a732fd9a425771cb3463d1cd143e5527acec751f32ba88f2df697aR72)() returns

Reply:
> Just making sure this implementation is solving the problem of not deleting deployments before running upgrade tests.

That's the main driver for bringing the [deployWindowsWorkloadAndTester()](https://github.com/openshift/windows-machine-config-operator/pull/645/commits/3551877c5aca09efea84533bb131bb105505ff78#diff-fc8ea8c7c1a732fd9a425771cb3463d1cd143e5527acec751f32ba88f2df697aL157) content here, to ensure deployments are running while running the upgrade test.